jax/_src/extend/ffi.py

```diff
@@ -140,3 +148,110 @@ def _ir_attribute(obj: Any) -> ir.Attribute:
     elif isinstance(mlir_type, ir.FloatType):
       return ir.FloatAttr.get(mlir_type, obj)
   raise TypeError(f"Unsupported attribute type: {type(obj)}")
+
+
+def ffi_call(
+    target_name: str,
+    result_shape_dtypes: DuckTypedArray | Sequence[DuckTypedArray],
```

Comment:
OOC why not jax.ShapeDtypeStruct as in pure_callback?

Reply:
`pure_callback` is typed to take [`Any` for `result_shape_dtypes`](https://github.com/google/jax/blob/a386af446caeb0eb908e9c72418b91b55408a573/jax/_src/callback.py#L261) because it accepts a pytree, but the leaves are implicitly expected to satisfy `DuckTypedArray`. In both cases, the docstrings say that `ShapeDtypeStruct` is useful for constructing this argument, but the only requirement in each case is that the leaves have `.shape` and `.dtype` properties. With this in mind, this seemed right to use this annotation here, rather than the stricter `ShapeDtypeStruct`!